cpp/include/rapidsmpf/coll/allreduce.hpp

```diff
@@ -0,0 +1,464 @@
+struct is_supported_reduce_op<T, DeviceReduceOp, DeviceReduceOp::PROD>
+    : is_supported_arithmetic_op<T> {};
+
+template <typename T>
+struct is_supported_reduce_op<T, DeviceReduceOp, DeviceReduceOp::MIN>
+    : is_supported_arithmetic_op<T> {};
+
+template <typename T>
+struct is_supported_reduce_op<T, DeviceReduceOp, DeviceReduceOp::MAX>
+    : is_supported_arithmetic_op<T> {};
+
+template <typename T, typename OpEnum, OpEnum Op>
+inline constexpr bool is_supported_reduce_op_v =
+    is_supported_reduce_op<T, OpEnum, Op>::value;
+
+/**
+ * @brief Create a host-based element-wise reduction operator for a given (T, Op).
+ *
+ * The operator assumes that the `PackedData::data` buffers reside in host memory
+ * and contain a contiguous array of @p T with identical sizes on all ranks.
+ */
+template <typename T, HostReduceOp Op>
+ReduceOperatorFunction make_host_reduce_operator_impl() {
+    static_assert(
+        is_supported_reduce_op_v<T, HostReduceOp, Op>,
+        "make_host_reduce_operator_impl called for unsupported (T, Op) combination"
+    );
+
+    auto const apply = [](PackedData& accum, PackedData&& incoming, auto&& op) {
+        RAPIDSMPF_EXPECTS(
+            accum.data && incoming.data,
+            "AllReduce reduction operator requires non-null data buffers"
+        );
+
+        auto* acc_buf = accum.data.get();
+        auto* in_buf = incoming.data.get();
+
+        auto const acc_nbytes = acc_buf->size;
+        auto const in_nbytes = in_buf->size;
+        RAPIDSMPF_EXPECTS(
+            acc_nbytes == in_nbytes,
+            "AllReduce reduction operator requires equal-sized buffers"
+        );
+
+        auto const nbytes = acc_nbytes;
+        RAPIDSMPF_EXPECTS(
+            nbytes % sizeof(T) == 0,
+            "AllReduce reduction operator requires buffer size to be a multiple of "
+            "sizeof(T)"
+        );
+
+        auto const count = nbytes / sizeof(T);
+
+        RAPIDSMPF_EXPECTS(
+            acc_buf->mem_type() == MemoryType::HOST
+                && in_buf->mem_type() == MemoryType::HOST,
+            "make_host_reduce_operator expects host memory"
+        );
+
+        auto* acc_bytes = acc_buf->exclusive_data_access();
+        auto* in_bytes = in_buf->exclusive_data_access();
+        auto* acc_ptr = reinterpret_cast<T*>(acc_bytes);
+        auto const* in_ptr = reinterpret_cast<T const*>(in_bytes);
+
+        for (std::size_t i = 0; i < count; ++i) {
+            acc_ptr[i] = op(acc_ptr[i], in_ptr[i]);
+        }
+
+        acc_buf->unlock();
+        in_buf->unlock();
+    };
+
+    if constexpr (Op == HostReduceOp::SUM) {
+        return [apply](PackedData& accum, PackedData&& incoming) {
+            if constexpr (std::is_same_v<T, bool>) {
+                apply(accum, std::move(incoming), [](T a, T b) {
+                    return static_cast<T>(a || b);
+                });
+            } else {
+                apply(accum, std::move(incoming), std::plus<T>{});
+            }
```

Comment:
I think `std::plus<bool>` behaves the same as logical OR

Reply:
Yes, I thought so too, just wanted to be verbose, but removed now with B.